rusoto/signature/src/sign.rs

```diff
@@ -0,0 +1,181 @@
+#[cfg(feature = "futures-01")]
+use crate::stream::ByteStream;
+#[cfg(feature = "futures-01")]
+use futures::stream::Stream;
+#[cfg(feature = "futures-01")]
+use hyper::{Body, Chunk};
+
+#[cfg(feature = "async-await")]
+use futures::stream::TryStreamExt;
+#[cfg(feature = "async-await")]
+use hyper_alpha::{Body, Chunk};
+
+use crate::credential::AwsCredentials;
+use crate::{Region, SignedRequest};
+use http::{Error, Request};
+use std::convert::TryInto;
+
+/// A trait to apply AWS' signatures. At the moment, this trait is only implemented for
+/// [http::Request] with Hyper's default body.
+pub trait Sign: Sized {
+    type Error;
+    /// Returns a copy of self, signed
+    fn sign(
+        self,
```

Comment:
why move over `&mut self`?

Reply:
The impact of this is if the sign fails you lose the request?

Reply:
You probably want to follow this signature https://docs.rs/http/0.1.19/http/request/struct.Builder.html#method.body

Reply:
The move is related to the need for the `SignedRequest` type to consume the http Request types body in order to compute the signature.

There isn't much a caller can do with the request if the signing fails. At least not currently.

Reply:
I think in that case it makes more sense to provide a specific aws body up front. The body can containa `&self` sign method and the body also contains  the whole body already as a `Bytes`. This way you can create multiple signed requests from it without needing to consume things.

You can even make a SigV4Body trait that contains `&self` sign method. I don't think you need any of the lazy body polling features here since there is no streaming support?

Reply:
Unless I misunderstood you, there is support for streaming in AWS Services, most notably in S3 and Kinesis.

Reply:
@davidbarsky then maybe I'm miss understanding how you would sign that body?

Reply:
That said, I think `http` 0.2 will move to move by value builders anyways so this may be a non issue.

Reply:
Yea, I'd say let's not try to optimize too much upfront until we have more use cases that demand it.